optimum/habana/peft/layer.py

```diff
@@ -0,0 +1,36 @@
+import math
+import warnings
+from typing import Any, List, Optional, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from transformers.pytorch_utils import Conv1D
+
+
+def GaudiLoraLayerLinearforward(self, x: torch.Tensor, *args: Any, **kwargs: Any) -> torch.Tensor:
+    #https://github.com/huggingface/peft/blob/4b02148af252c17e36b0a4b995f9e8519806fbb5/src/peft/tuners/lora/layer.py#L354C1-L376C22
+    #only differences are avoiding inplace update of "result" to prevent error from torch Dynamo in torch.compile mode of execution
+    #and replacing self.base_layer by self._linear
+    previous_dtype = x.dtype
+
+    if self.disable_adapters:
+        if self.merged:
+            self.unmerge()
+        result = self._linear(x, *args, **kwargs)
+    elif self.merged:
+        result = self._linear(x, *args, **kwargs)
+    else:
+        result = self._linear(x, *args, **kwargs)
+        for active_adapter in self.active_adapters:
+            if active_adapter not in self.lora_A.keys():
+                continue
+            lora_A = self.lora_A[active_adapter]
+            lora_B = self.lora_B[active_adapter]
+            dropout = self.lora_dropout[active_adapter]
+            scaling = self.scaling[active_adapter]
+            x = x.to(lora_A.weight.dtype)
+            result = result.clone() + lora_B(lora_A(dropout(x))) * scaling
```

Comment:
```suggestion
            result = result + lora_B(lora_A(dropout(x))) * scaling
```
I think we can remove `.clone()` and it will still work.
It seems that in Torch, `a += b` is inplace but not `a = a + b`: https://discuss.pytorch.org/t/what-is-in-place-operation/16244

Reply:
Currently we support FSDP only with torch_compile mode, in this mode we see following error if do not clone.

**RuntimeError: Output 0 of CompiledFunctionBackward is a view and is being modified inplace. This view was created inside a custom Function (or because an input was returned as-is) and the autograd logic to handle view+inplace would override the custom backward associated with the custom Function, leading to incorrect gradients. This behavior is forbidden. You can fix this by cloning the output of the custom Function.**

Corresponding issue on PyTorch captured here https://github.com/pytorch/pytorch/issues/90507

Reply:
Ah okay, I didn't know the behavior is different with torch compile :+1: